nixos/modules/services/networking/doh-server.md

```diff
@@ -0,0 +1,64 @@
+# DNS-over-HTTPS Server {#module-service-doh-server}
+
+[DNS-over-HTTPS](https://github.com/m13253/dns-over-https) is a high performance DNS over HTTPS client & server. This module enables its server part (`doh-server`).
+
+## Quick Start {#module-service-doh-server-quick-start}
+
+`doh-server` can be used as a standalone https web server (with ssl cert and key speicified) or with other reverse proxy services (such as `nginx` or `candy`). The latter approach is recommended as `doh-server` does not do OCSP Stapling.
+
+Setup a standalone instance:
+
+```nix
+services.doh-server = {
+  enable = true;
+  settings = {
+    listen = [ ":443" ];
+    upstream = [ "udp:1.1.1.1:53" ];
+    cert = "/path/to/cert.crt";
+    key = "/path/to/key.key";
+  };
+};
+```
+
+Setup with Nginx + ACME (recommanded):
+
+```nix
+services.doh-server = {
+  enable = true;
+  settings = {
+    listen = [ ":8053" ];
+    upstream = [ "udp:1.1.1.1:53" ];
+  };
+};
+
+services.nginx = {
+  enable = true;
+  virtualHosts."doh.example.com" = {
+    enableACME = true;
+    forceSSL = true;
+    http2 = true;
+    locations."/".return = 404;
+    locations."/dns-query".extraConfig = ''
```

Comment:
Most of these can be replaced by `recommendedProxySettings`, `proxyWebsockets` and `proxyPass`